source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1119,42 +1171,219 @@ void ClusterManagerImpl::postThreadLocalClusterUpdate(ClusterManagerCluster& cm_
+            per_priority.locality_weights_, per_priority.hosts_added_, per_priority.hosts_removed_,
+            per_priority.weighted_priority_health_, per_priority.overprovisioning_factor_, map);
+      }
+
+      if (new_cluster != nullptr) {
+        ThreadLocalClusterCommand command = [&new_cluster]() -> ThreadLocalCluster& {
+          return *new_cluster;
+        };
+        for (auto cb_it = cluster_manager->update_callbacks_.begin();
+             cb_it != cluster_manager->update_callbacks_.end();) {
+          // The current callback may remove itself from the list, so a handle for
+          // the next item is fetched before calling the callback.
+          auto curr_cb_it = cb_it;
+          ++cb_it;
+          (*curr_cb_it)->onClusterAddOrUpdate(info->name(), command);
+        }
       }
     }
   });
 }
 
+ClusterManagerImpl::ClusterInitializationObjectConstSharedPtr
+ClusterManagerImpl::addOrUpdateClusterInitializationObjectIfSupported(
+    const ThreadLocalClusterUpdateParams& params, ClusterInfoConstSharedPtr cluster_info,
+    LoadBalancerFactorySharedPtr load_balancer_factory, HostMapConstSharedPtr map) {
+  if (!deferralIsSupportedForCluster(cluster_info)) {
+    return nullptr;
+  }
+
+  const std::string& cluster_name = cluster_info->name();
+  auto entry = cluster_initialization_map_.find(cluster_name);
+  // TODO(kbaichoo): if EDS can be configured via cluster_type() then modify the
+  // merging logic below.
+  const bool should_merge_with_prior_cluster =
+      entry != cluster_initialization_map_.end() &&
+      entry->second->cluster_info_->type() == cluster_info->type() &&
+      cluster_info->type() == envoy::config::cluster::v3::Cluster::EDS;
+
+  if (should_merge_with_prior_cluster) {
+    // We need to copy from an existing Cluster Initialization Object. In
+    // particular, only update the params with changed priority.
+    auto new_initialization_object = std::make_shared<ClusterInitializationObject>(
+        entry->second->per_priority_state_, params, std::move(cluster_info), load_balancer_factory,
+        map);
+    cluster_initialization_map_[cluster_name] = new_initialization_object;
+    return new_initialization_object;
+  } else {
+    // We need to create a fresh Cluster Initialization Object.
+    auto new_initialization_object = std::make_shared<ClusterInitializationObject>(
+        params, std::move(cluster_info), load_balancer_factory, map);
+    cluster_initialization_map_[cluster_name] = new_initialization_object;
+    return new_initialization_object;
+  }
+}
+
+ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry*
+ClusterManagerImpl::ThreadLocalClusterManagerImpl::initializeClusterInlineIfExists(
+    absl::string_view cluster) {
+  auto entry = thread_local_deferred_clusters_.find(cluster);
+  if (entry == thread_local_deferred_clusters_.end()) {
+    // Unknown cluster.
+    return nullptr;
+  }
+
```

Comment:
Should the cluster_name be removed from `thread_local_deferred_clusters_` (maintaining an invariant that a cluster is either deferred or not)?

Reply:
good idea, done